### PR TITLE
New metrics to return Inventory table with quantity in units of kg/GWe

### PR DESCRIPTION
**Pull request by louishartono**
Made some changes to metrics.py. Added a new metrics called "InventoryQuantityPerGWe" to return the inventory table with quantity in units of kg/GWe instead of kg and created some tests to check the expected output as well. Feel free to give feedback or suggestions :).

### CODE REVIEW

File: cymetric/metrics.py [hunk 1]
```diff
@@ -441,3 +441,58 @@ def timelist(info):
...
+    ('AgentId', ts.INT),
+    ('Prototype', ts.STRING),
+    ('Time', ts.INT),
+    ('InventoryName', ts.STRING),
+    ('NucId', ts.INT),
+    ('Quantity', ts.DOUBLE)
+    ]
+
+@metric(name='', depends=_invdeps, schema=_invschema)
+def (inv):
```
Comment (bam241) [inline cymetric/metrics.py]: you will need to specify a name for you metrics

File: cymetric/metrics.py [hunk 2]
```diff
@@ -441,3 +441,58 @@ def timelist(info):
 
 del _tldeps, _tlschema
 
+# Quantity per GigaWattElectric in Inventory [kg/GWe]
+_invdeps = ['']
```
Comment (bam241) [inline cymetric/metrics.py]: you will need some here:
dependencies are the existing tables (or metrics) you rely on to build the new one

File: cymetric/metrics.py [hunk 3]
```diff
@@ -441,3 +441,58 @@ def timelist(info):
...
+    inv = tools.raw_to_series(expinv,
+                              ['SimId', 'AgentId', 'Prototype', 'Time', 'InventoryName', 'NucId'],
+                              'Quantity')
+    inv = inv.groupby(level=inv_index).sum()
+    inv.name = 'Quantity'
+    rtn = inv.reset_index()
+    return rtn
+
+# Mass per GigaWattElectric in Transaction [kg/GWe]
+_transdeps = ['']
```
Comment (bam241) [inline cymetric/metrics.py]: same here some dependencies are required

File: cymetric/metrics.py [hunk 4]
```diff
@@ -441,3 +441,58 @@ def timelist(info):
...
+def (inv):
+    """Returns quantity per GWe in the inventory table
+    """
+    inv_index = ['SimId', 'AgentId', 'Prototype', 'Time', 'InventoryName', 'NucId']
+    inv = tools.raw_to_series(expinv,
+                              ['SimId', 'AgentId', 'Prototype', 'Time', 'InventoryName', 'NucId'],
+                              'Quantity')
+    inv = inv.groupby(level=inv_index).sum()
+    inv.name = 'Quantity'
+    rtn = inv.reset_index()
```
Comment (bam241) [inline cymetric/metrics.py]: inside this method, you will need to implement some version of the algorithm you develop to compute the inventory per GWe

File: cymetric/metrics.py [hunk 5]
```diff
@@ -441,3 +441,58 @@ def timelist(info):
...
+@metric(name='', depends=_transdeps, schema=_transschema)
+def (mats, tranacts):
+    """Returns mass per GWe in the transaction table
+    """
+    trans_index = ['SimId', 'ResourceId', 'NucId', 'Mass',
+            'ReceiverID', 'SenderId', 'SenderPrototype', 'TransactionID', 'Commodity', 'Time']
+    trans = pd.merge(mats, tranacts, on=['SimId', 'ResourceId'], how='inner')
+    trans = trans.set_index(trans_index)
+    trans = trans.groupby(level=trans_index)['Mass'].sum()
+    trans.name = 'Quantity'
```
Comment (bam241) [inline cymetric/metrics.py]: Same here the code to compute the actual value will need to be implemented.

File: cymetric/metrics.py [hunk 6]
```diff
@@ -441,3 +441,58 @@ def timelist(info):
...
+    ('ReceiverPrototype', ts. STRING)
+    ('SenderId', ts.INT),
+    ('SenderPrototype', ts. STRING)
+    ('TransactionId', ts.INT),
+    ('Commodity', ts.STRING),
+    ('Time', ts.INT),
+    ]
+
+@metric(name='', depends=_transdeps, schema=_transschema)
+def (mats, tranacts):
```
Comment (bam241) [inline cymetric/metrics.py]: metrics is missing its name.

File: cymetric/metrics.py (hunk 2 above)
Comment (bam241) [reply cymetric/metrics.py]: so I am guessing, `Inventory`, `Power`... (I am not sure this is there exact names...)

**Comment by bam241** [comment]
@louishartono you should also take the time to explain what this is about in the description of the PR, so other people get some content :)

**Comment by bam241** [comment]
@louishartono added the `[WIP]` tag to allow other developers to know that this work is still ongoing and not ready to be included in the main repo.

File: cymetric/metrics.py [hunk 7]
```diff
@@ -441,3 +441,86 @@ def timelist(info):
...
+			     'Time': expinv.Time, 
+		             'InventoryName': expinv.InventoryName,
+	 		     'NucId': expinv.NucId,
+                             'Quantity': expinv.Quantity},
+	             columns=['SimId','AgentId','Time','InventoryName','NucId','Quantity'])    
+    for x in range(len(inv)):
+        for y in range(len(df1)):
+            time = inv.Time[x]
+            if df1.Time[y]==time and df1.SimId[y]==inv.SimId[x]:
+                inv.Quantity[x] == inv.Quantity[x]//df1.Value[y]
```
Comment (bam241) [inline cymetric/metrics.py]: why the `//` and not a single `/` ?

File: tests/test_metrics.py [hunk 8]
```diff
@@ -367,3 +367,32 @@ def test_timelist():
...
+    tsp = pd.DataFrame(np.array([
+        (UUID('f22f2281-2464-420a-8325-37320fd418f8'), 1, 0, 100),
+        (UUID('f22f2281-2464-420a-8325-37320fd418f8'), 2, 0, 200),
+        (UUID('f22f2281-2464-420a-8325-37320fd418f8'), 1, 1, 100),
+        ], dtype=ensure_dt_bytes([
+                ('SimId', 'O'), ('AgentId', '<i8'), ('Time', '<i8'),
+                ('Value', '<f8')]))
+        )
+    inv = pd.DataFrame(np.array([
+        (UUID('f22f2281-2464-420a-8325-37320fd418f8'), 1, 0, 'core', 922350000, 300),
```
Comment (bam241) [inline tests/test_metrics.py]:
```suggestion
        (UUID('f22f2281-2464-420a-8325-37320fd418f8'), 1, 0, 'usedfuel', 922350000, 300),
```

File: tests/test_metrics.py [hunk 9]
```diff
@@ -367,3 +367,32 @@ def test_timelist():
 
 if __name__ == "__main__":
     nose.runmodule()
+
+def test_inventory_quantity_per_gwe():
+    exp = pd.DataFrame(np.array([
+        (UUID('f22f2281-2464-420a-8325-37320fd418f8'), 1, 0, 'core', 922350000, 1.0),
```
Comment (bam241) [inline tests/test_metrics.py]:
```suggestion
        (UUID('f22f2281-2464-420a-8325-37320fd418f8'), 1, 0, 'usedfuel', 922350000, 1.0),
```

File: tests/test_metrics.py [hunk 10]
```diff
@@ -367,3 +367,32 @@ def test_timelist():
...
+        (UUID('f22f2281-2464-420a-8325-37320fd418f8'), 1, 0, 'core', 922350000, 2.0),
+        (UUID('f22f2281-2464-420a-8325-37320fd418f8'), 2, 1, 'core', 922350000, 2.0),
+        ], dtype=ensure_dt_bytes([
+                ('SimId', 'O'), ('AgentId', '<i8'), ('Time', '<i8'),
+                ('InventoryName', 'O'), ('NucId', '<i8'), ('Quantity', '<f8')]))
+        )
+    tsp = pd.DataFrame(np.array([
```
Comment (bam241) [inline tests/test_metrics.py]: maybe add some comment to explain what are `tsp`/`exp`/`inv`

File: tests/test_metrics.py [hunk 11]
```diff
@@ -367,3 +367,32 @@ def test_timelist():
 
 if __name__ == "__main__":
     nose.runmodule()
```
Comment (bam241) [inline tests/test_metrics.py]: those lines are supposed to be at the end of the file.

**Comment by bam241** [comment]
looking good for me.
I'll let @gonuke merge.

**Comment by gonuke** [comment]
Thanks for this first contribution @louishartono !

**Pull request merged**